### ISSUE TITLE
Add a SliceListing that lists file in a directory

### ISSUE THREAD

**Issue by victornoel** [artipie/http#158]
This is needed for artipie/maven-adapter#87.

We need a slice that, given a key in the storage, will return a list of the files present in that directory in the form of a simple HTML listing.

We will also maybe need a `SliceRoute` that allow to dispatch requests depending on the fact a key is a directory or a file, for example to be able to say that GET requests get either dispatcher to `SliceDownload` or `SliceListing`. See https://github.com/artipie/http/issues/158#issuecomment-633629631 below for the proposed design.

**Comment by 0crat** [comment]
@g4s8[/z](https://www.0crat.com/u/g4s8) please, pay attention to this issue

**Comment by g4s8** [comment]
@victornoel we don't actually have directories or files, we use `Storage` abstraction to hide persistence implementation details. E.g. AWS S3 storage is an **object storage**, it means that it **imitates** directory behavior by using key prefixes, but S3 doesn't have directories or files, and we are not able to check it.
The difference is:
 - file system: `foo/bar/file.text` - hierarchical structure, where `file.txt` is a file, `bar` is a directory of `file.txt`, `foo` is a directory of `bar`
 - S3 storage (or any other object storage): `foo/bar/file.txt` - a key name (string) which is used to find binary data in storage

**Comment by victornoel** [comment]
@g4s8 ah I understand, so basically, if `storage.list(key)` returns one element that is equal to the key, we consider it's a file, and if it returns multiple elements prefixed by the key then we can consider it a directory for example?

**Comment by g4s8** [comment]
> @g4s8 ah I understand, so basically, if `storage.list(key)` returns one element that is equal to the key, we consider it's a file, and if it returns multiple elements prefixed by the key then we can consider it a directory for example?

@victornoel yes, maybe we can work around it somehow, e.g. by checking for `/` char `list` result. In case of S3 (and file system storage too for consistency) it returns all keys in storage filtered by prefix. E.g. if you run `storage.list("maven")` it will return:
```
maven/com/artipie/http/1.0/http-1.0.jar
maven/com/artipie/http/1.0/http-1.0-sources.jar
maven/com/artipie/http/2.0/http-2.0.jar
maven/com/artipie/http/2.0/http-2.0-sources.jar
maven/com/artipie/http/maven-metadata.xml
```

**Comment by victornoel** [comment]
@g4s8 but then it's not completely clear how, with the current implementation of `SliceDownload`, we are going to discriminate between the two cases, maybe `SliceDownload` could have an optional constructor argument to delegate to another `Slice` if the file is not found for example? (and the default behaviour would be a 404 `Slice`...).

**Comment by victornoel** [comment]
@g4s8 and actually, SliceListing would also behave the same, so that we can write:
```java
new SliceRoute.Path(
  new RtRule.ByMethod(RqMethod.GET),
  new SliceAuth(
    new SliceDownload(
        storage,
        new SliceListing(
          storage,
          new Slice404()
        )
     ),
     new Permission.ByName("download", perms),
     users
  )
)
```

Or `Slice` could return an `Optional` and have another abstraction responsible of chaining `Slices` in case they return empty?

**Comment by g4s8** [comment]
@victornoel I don't really like to include `Listing` as a part of `Download` or wise versa, `Optional` is not a valid scenario for `Slice`, since it doesn't return a data object, but `Response` which can send itself to connection. What do you think about this case:
```java
new SliceRoute.Path(
  new RtRule.ByMethod(RqMethod.GET),
  new SliceRoute.Path(
    new IfExistsRule(), // checks if key from path exists in storage
    new SliceDownload(storage)
  ),
  new SliceRoute.Path(
    RtRule.FALLBACK,
    SliceListing(storage)
  )
)
```

**Comment by victornoel** [comment]
@g4s8 sorry for the delay, that seems good yes, I will update the description

**Comment by g4s8** [comment]
@0crat assign @paulodamaso

**Comment by 0crat** [comment]
> @0crat assign @paulodamaso  ([here](https://github.com/artipie/http/issues/158#issuecomment-636799389))

@g4s8 The job [#158](https://github.com/artipie/http/issues/158) assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/999eab5f-5985-4ecd-a0be-dac0a26a69c0); the budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9); if the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html); there will be a monetary reward for this job

**Comment by paulodamaso** [comment]
@0crat wait for https://github.com/artipie/http/pull/172

**Comment by g4s8** [comment]
Changes from https://github.com/artipie/artipie/pull/548 PR can be used for keys formatting

**Comment by g4s8** [comment]
@dgarus please check https://github.com/artipie/files-adapter/issues/77 issue, it was partially fixed by https://github.com/artipie/files-adapter/pull/89 - we need some base implementation for listing slice to include into different adapters and customize output format

**Comment by 0pdd** [comment]
@victornoel the puzzle [#unknown]() is still not solved; solved: [#426](https://github.com/artipie/http/issues/426).
